### ISSUE TITLE
java.lang.NoSuchMethodError: org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.fromRow

### ISSUE THREAD

**Issue by ar-ms** [GoogleCloudDataproc/spark-bigquery-connector#273]
Hello Folks!

I have the following issue when I'm trying to stream data to BQ, the normal write does work.
I'm using the following image PREVIEW 2.0 (Ubuntu 18.04 LTS, Hadoop 3.2, Spark 3.0), and executing the following command inside the instance:

```pyspark --packages="org.apache.spark:spark-streaming_2.12:3.0.1,io.delta:delta-core_2.12:0.7.0" --jars=gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar```

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2060)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2009)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2008)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2008)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:973)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:973)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:973)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2240)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2189)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2178)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:775)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2099)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:195)
	... 45 more
Caused by: java.lang.NoSuchMethodError: org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.fromRow(Lorg/apache/spark/sql/catalyst/InternalRow;)Ljava/lang/Object;
	at com.google.cloud.spark.bigquery.BigQueryStreamWriter$.$anonfun$writeBatch$2(BigQueryStreamWriter.scala:46)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeTask(FileFormatWriter.scala:260)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$write$15(FileFormatWriter.scala:205)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:127)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Comment by davidrabinowitz** [comment]
Thanks, we will check it out.

**Comment by ar-ms** [comment]
Thanks @davidrabinowitz  for pointing out, but the normal write is working on my side, but the writeStream does not..

The steam part that fails:
```
df = spark.readStream.format("delta").load("gs://XXXX.delta")

df.writeStream \
  .format("bigquery") \
  .option("persistentGcsBucket","gq_loader_tmp") \
  .option("checkpointLocation", "gs://YYY/yyy") \
  .option("table", "ZZZZ").start().awaitTermination()
``` 

the batch part that works:
```
df_phone_calls = spark.read.format("delta").load("gs://XXX.delta")

df_phone_calls.write \
  .format("bigquery") \
  .mode("overwrite") \
  .option("persistentGcsBucket","gq_loader_tmp") \
  .option("checkpointLocation", "gs://YYY/yyy") \
  .option("table", "ZZZ").save()
```

I tried with `persistentGcsBucket` and with `temporaryGcsBucket` and both works for the batch write. Neitheir work with `writeStream`

**Comment by ar-ms** [comment]
I just tried with Spark 2.4.7 with exactly the same code and it works perfectly !

**Comment by gabrielcortes-gavb** [comment]
This issue is still happening for Spark 3.1.1. I had today the exaclty same problem and needed to downgrade my spark version to 2.4.7.

**Comment by andr9184** [comment]
The issue still happening for Spark 3.1.1 and writeStream to the BigQuery. Only working as mentioned before with the Spark 2.4.7.

**Comment by andr0267** [comment]
In Spark 3.x method `ExpressionEncoder.fromRow `was removed (https://github.com/apache/spark/commit/e7fef70fbbea08a38316abdaa9445123bb8c39e2)

That means that current code of BigQueryStreamWriter.scala needs to be modified to support Spark 3.x into something like this:
` val expressionEncoder: ExpressionEncoder[Row] = RowEncoder(schema).resolveAndBind()
    val deserializer = expressionEncoder.createDeserializer()
    val rowRdd: RDD[Row] = data.queryExecution.toRdd.mapPartitions(iter =>
      iter.map(internalRow =>
        deserializer.apply(internalRow)
      )
    )` 
however then it means that we can't use one version of this library for both Spark 2.x and 3.x

**Comment by himanshukohli09** [comment]
This issue is now resolved, see PR #396
